### ISSUE TITLE
Intermediate states ignored?

### ISSUE THREAD

**Issue by haoyangz** [kuz/caffe-with-spearmint#2]
It seems to be that current code ignores intermediate states in each run (i.e. each parameter setting), and only takes the accuracy/loss at the last iteration as the metric for the that run. But in practice, due to overfitting and other issue, the model trained till last iteration is usually not the best one to use.  Why not just take the best accuracy/loss in each run as its accuracy/loss, and compare different runs on that?

**Comment by kuz** [comment]
My approach is to set the number of iterations to be low (to cover, say, 3 epochs) during the parameters search. Overfitting is among the things I would like to avoid, so if my network at 100 iterations performs well, but at 300 iterations it already overfits I would like to discard the set of parameters which led to that. If I would use best accuracy/loss then I will not notice overfitting.

Of course if you train the network to "full" extent during each run then best accuracy/loss is reasonable. The only problem there is that it will take a lot of time.

But I agree that for smaller datasets and/or smaller parameter search spaces it makes sense. I will add it as a parameter. Something like `--optimize accuracy --optimizewrt last/max`

**Comment by haoyangz** [comment]
Thank you kuz for the change! If I understand it correctly, now using --best we can get the loss of the best iteration, but it will still save the model state of the last iteration. Is this true? If so, I think it makes more sense to save the model with the best loss instead?
